smart-contracts/migrations/2_deploy_unlock.js

```diff
@@ -1,21 +1,23 @@
-const Unlock = artifacts.require('./Unlock.sol')
-const shell = require('shelljs')
+// Load zos scripts and truffle wrapper function
+const { scripts, ConfigVariablesInitializer } = require('zos')
+const { add, push, create } = scripts
 
-module.exports = function deployProxy (deployer, network, accounts) {
-  const unlockOwner = accounts[0]
-  const proxyAdmin = accounts[9]
+module.exports = function (deployer, networkName, accounts) {
+  deployer.then(async () => {
+    const { network, txParams } = await ConfigVariablesInitializer.initNetworkConfiguration({ network: networkName, from: accounts[1] })
+    txParams.gas = 4000000
+    const options = { network, txParams }
+    // Register v0 of MyContract in the zos project
+    add({ contractsData: [{ name: 'Unlock', alias: 'Unlock' }] })
 
-  deployer.then(() => {
-    if (network === 'test' || network === 'development') {
-      return deployer.deploy(Unlock, unlockOwner)
-    } else {
-      if (
-        shell.exec(
-          `zos push --reset --network ${network} --from ${proxyAdmin}`
-        ).code !== 0
-      ) {
-        throw new Error('Migration failed')
-      }
-    }
+    // Push implementation contracts to the network
+    await push(options)
+
+    // Create an instance of MyContract, setting initial value to 42
+    await create(Object.assign({
```

Comment:
I wonder about this part. I know it follows the zos example, but I believe there's a good reason to keep the deployment of unlock and the creation of the proxy in separate scripts. 
If this script fails at the proxy deployment (`create`) step, the Unlock contract has already been deployed, but the number of ther last successful migration in `Migrations.sol` won't be updated. So , next time we run `truffle deploy`, the Unlock contract will be deployed again by the `push` step. 
I feel like these migrations scripts should be atomic and do one thing each to avoid duplicating steps unintentionally.

Reply:
If any line in the migration step fails, the entire file will run again on next migration - i.e. it should fix itself.  Are you seeing different behavior?

Reply:
Well, if the `push` step is successful, Unlock.sol is deployed. But if the script fails after that, the entire file will run again as you say, and will deploy another instance of Unlock.sol. Not a big deal in development, but if we want to use migrations consistently (including for upgrades on rinkeby/mainnet) it's probably better for these scripts to be atomic steps, ie: the script either deploys unlock or it doesn't, then the next one either creates the proxy or doesn't, etc...
Now if there's no intention of using migrations for upgrades, then this isn't as important. But, as all of our actual deployments and upgrades have been done manually so far, there's definitely room to improve on that by scripting the process.

Reply:
I see... k makes sense, will change

Reply:
Done